### ISSUE TITLE
Track_grad_norm only tracks the parameters of the last optimizer defined

### ISSUE THREAD

**Issue by MikkelAntonsen** [PyTorchLightning/pytorch-lightning#1527]
## 🐛 Bug

When you enable track_grad_norm in Trainer you expect it to track the grad of all the parameters defined in your lightning module. It seems like it only tracks the parameters for the last optimizer defined in `def configure_optimizers()`.  

### To Reproduce

1. Create a pytorch lightning module with more than one optimizer
2. Enable logging to tensorboard and track_grad_norm in the trainer
3. Train model and observe only gradients for the last optimzer is tracked
4. Delete all tensorboard data and restart tensorboard
5. Train a model with the optimizers returned in the different order. Observe which gradients are tracked again

#### Code sample

```
import pytorch_lightning as pl
from torch.utils.data import TensorDataset, DataLoader
from pytorch_lightning.loggers.tensorboard import TensorBoardLogger
from torch.optim import SGD
import torch.nn as nn
import torch

class MWENet(pl.LightningModule):
    def __init__(self):
        super(MWENet, self).__init__()
        
        self.first = nn.Conv2d(1, 1, 3)
        self.second = nn.Conv2d(1, 1, 3)
        self.loss = nn.L1Loss()
        
    def train_dataloader(self):
        xs, ys = torch.zeros(16, 1, 10, 10), torch.ones(16 ,1, 6, 6)
        ds = TensorDataset(xs, ys)
        return DataLoader(ds)
    
    def forward(self, xs):
        out = self.first(xs)
        out = self.second(out)
        return out
    
    def configure_optimizers(self):
        first = SGD(self.first.parameters(), lr=0.01)
        second = SGD(self.second.parameters(), lr=0.01)
        return [second, first]
    
    def training_step(self, batch, batch_idx, optimizer_idx):
        xs, ys = batch
        out = self.forward(xs)
        return {'loss': self.loss(out, ys)}

net = MWENet()
logger = TensorBoardLogger('somedir', name='testing')
trainer = pl.Trainer(track_grad_norm=2,
                     logger = logger)
trainer.fit(net)
```

### Expected behavior

I expect that a pytorch module will report the gradient of all parameters when track_grad_norm is enabled.

### Environment

```
cuda:
	GPU:
		GeForce RTX 2080 Ti
	available:           True
	version:             10.1
packages:
	numpy:               1.18.1
	pyTorch_debug:       False
	pyTorch_version:     1.4.0
	pytorch-lightning:   0.7.3
	tensorboard:         2.2.1
	tqdm:                4.45.0
system:
	OS:                  Windows
	architecture:
		64bit
		WindowsPE
	processor:           Intel64 Family 6 Model 158 Stepping 12, GenuineIntel
	python:              3.7.7
	version:             10.0.18362
```

**Comment by windj007** [comment]
Any progress on this issue? I face this too.

**Comment by MartFire** [comment]
Still facing the issue when training a GAN with PL 1.0.7, either the generator or discriminator has all tracked gradients to zero in TensorBoard

**Comment by MartFire** [comment]
A workaround is to implement `on_after_backward` to have access to the gradients and log them in your logger
https://pytorch-lightning.readthedocs.io/en/stable/lightning_module.html#on-after-backward

**Comment by MartFire** [comment]
I think I found the problem. I implemented `on_after_backward` following the example in https://pytorch-lightning.readthedocs.io/en/stable/lightning_module.html#on-after-backward.
````
def on_after_backward(self):
    # example to inspect gradient information in tensorboard
    if self.trainer.global_step % 25 == 0:  # don't make the tf file huge
        params = self.state_dict()
        for k, v in params.items():
            grads = v
            name = k
            self.logger.experiment.add_histogram(tag=name, values=grads,
                                                 global_step=self.trainer.global_step)
````
First of all I think that example is wrong because `v` is not the gradient but the weight of the layer. A better example is
````
    def on_after_backward(self):
        # example to inspect gradient information in tensorboard
        if self.trainer.global_step % 25 == 0:  # don't make the tf file huge
        for k, v in zip(self.state_dict().keys(), self.parameters()):
            name = k
            # During the first step, when using multiple optimizers, gradients of parameters associated to an optimizer are equal
            # None as long as the optimizer.step() as not been called
            if v.grad is None:
                grads = 0
            else:         
                grads = v.grad           
            self.logger.experiment.add_scalar("grad_norm_2_" + name, torch.linalg.norm(grads),
                                                                  global_step=self.trainer.global_step)
````
But this example won't even work because `on_after_backward` is called for each optimizer. During one call, I have noticed that gradients of other optimizers are equal to zero. The last call correspond to the last optimizer returned by `configure_optimizers`. During this call all gradients of other optimizers are equal to zero

With this observation; I looked up Pytorch Lightning source code of https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_loop.py

In the function `run_training_batch` there is a loop over all the optimizers with the instruction

`grad_norm_dic = self.run_batch_backward_pass(split_batch, batch_idx, opt_idx, optimizer)`
so `grad_norm_dic` is overwrited at each iteration of the loop meaning that the returned `grad_norm_dic` is the one calculated for the last optimizer

The source code of `run_batch_backward_pass` is
````
def run_batch_backward_pass(self, split_batch, batch_idx, opt_idx, optimizer):
        # ------------------
        # GRAD NORMS
        # ------------------
        # track gradient norms when requested
        grad_norm_dic = {}
        if batch_idx % self.row_log_interval == 0:
            if float(self.track_grad_norm) > 0:
                model = self.get_model()
                grad_norm_dic = model.grad_norm(
                    self.track_grad_norm)

        # ------------------
        # CLIP GRADS
        # ------------------
        if self.amp_backend == AMPType.NATIVE and not self.use_tpu:
            self.scaler.unscale_(optimizer)
        self.clip_gradients(optimizer)

        # ------------------
        # .STEP + ZERO_GRAD
        # ------------------
        self.call_optimizer_step(optimizer, opt_idx, batch_idx, split_batch)

        return grad_norm_dic
````
We can see that `grad_norm_dic` is computed over all the parameters of the model not only the ones associated to the optimizer. That why I think that all the gradients are equal to zero except the ones associated to the optimizer.

`run_batch_backward_pass(self, split_batch, batch_idx, opt_idx, optimizer)` should only compute the gradients of the optimizer and its output should be inserted in a global dictionary instead of overwriting the existing one.

**Comment by MikkelAntonsen** [comment]
> run_batch_backward_pass(self, split_batch, batch_idx, opt_idx, optimizer) should only compute the gradients of the optimizer and its output should be inserted in a global dictionary instead of overwriting the existing one.

[This is mentioned as a todo ](https://github.com/PyTorchLightning/pytorch-lightning/blob/217650320e376f4dadd1c7b8c034ec55dee60a23/pytorch_lightning/trainer/training_loop.py#L712)

Where do you find `run_batch_backward_pass`? I can't find any reference to it.

**Comment by MartFire** [comment]
Sorry I was looking at an old version xD
Now the tracked norm is calculated with
`grad_norm_dic = self._cur_grad_norm_dict` 
in the optimizers loop so the overwriting still occurs
````

**Comment by edenlightning** [comment]
@ MikkelAntonsen any progress on this?

**Comment by MikkelAntonsen** [comment]
@edenlightning I've tried to implement a fix+some extra functionality. I'm awaiting feedback from @tchaton on what needs to be improved/fixed/done differently.

**Comment by tchaton** [comment]
Hey @MikkelAntonsen,

I added some comments on your PR. Do you need me to help you ?

Best regards,
T.C

**Comment by MikkelAntonsen** [comment]
Yeah, I've tried to fix what you've mentioned. If you see anything that could be better, I'd be all ears. There are a lot of failing tests in places I haven't touched (on purpose at least), is that normal?

**Comment by MikkelAntonsen** [comment]
Hey, I'm interested in trying to tie this branch up and I believe it shouldn't be too much work with some guidance, @tchaton!:)
